text/0000-network-spec.md

```diff
@@ -0,0 +1,160 @@
+
+This document should serve as reference for all the clients to implement networking layer.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## System description
+
+Our system is design on the random sampling of peers and cryptohraphic routing.
+Each node has ID that is ED25519 public key derived from secret key known only to this node.
+Additionally, active participants, such as block producers and other nodes that want to link their account information to the peer information have a string `account_id` and appropriate `public_key` which is associated with this account. Accounts maintained by blockchain itself, and can not be forged without breaking the blockchain itself.
+
+### Storing network information
+
+A client stores information about other nodes in three data structures. `known_peers` a long-term database, which is stored on disk and persists between reboots. And two short-term databases `active_peers` and `routing` that start empty.
+
+`known_peers`. The `known_peers` is stored on disk and contains information about each node that client has ever connected or recevied gossip about. There is no limit to the size of the `known_peers`.
+Each entry of `known_peers` is tuple of `<node ID, IP address, TCP port, last_seen, status>`. Where `last_seen` is a timestamp when we received last message from this peer. If this was received from gossip, `last_seen` is 0. Status contains if this peer is currently unknown, connected, disconnected or banned.
+
+`active_peers`. The `active_peers` maintains relevant information about currently connected peers. Such as their `<node ID, IP address, TCP port, last_seen, socket>`.
+
+`routing_table`. The `routing_table` contains for `account_id` a list of routes that are rooted in this node's peers.
+
+### Bootstrapping
+
+When node starts it have two lists:
+ - Bootstrap nodes, nodes that are well known in the eco-system as nodes that provide information about the network.
+  - Known peers, peers that this node has connected before in previous sessions and stored on disk to persist between runs. This list can be empty if this is first time to start the node.
+
+Both of these lists are merged and node starts by connecting to `max_peers` random peers from this list.
+If there is not enough peers in the lists, node periodically tries to connect to newly discovered peers with a exponential back off timeout (in case network is smaller than K in the first place) until it switches to *regular* mode. *Regular* mode means that node periodically checks if there is not enough peers (due to disconnects) and conncects to new peers.
+
+Additionally, to prevent various eclipse attacks where attacker forces node to have all incoming connection to them, we require node to have at least `min_outgoing_peers` outgoing connections.
+
+In pseudo code, this would look like:
+```
+  peers = bootstrap & known_peers - banned_peers
```

Comment:
Shouldn't this be `|` (union) instead of `&` (intersection)?
`peers = (bootstrap | known_peers) - banned_peers`